program/src/processor/slash.rs

```diff
@@ -11,14 +27,257 @@ use crate::instruction::accounts::{Context, SlashAccounts};
+    }
+
+    require!(
+        ctx.accounts.slash_authority.is_signer,
+        ProgramError::MissingRequiredSignature,
+        "stake authority",
+    );
+
+    // vault authority
+    // - derivation must match
+
+    let signer_bump = [config.vault_authority_bump];
+    let derivation = create_vault_pda(ctx.accounts.config.key, &signer_bump, program_id)?;
+
+    require!(
+        ctx.accounts.vault_authority.key == &derivation,
+        StakeError::InvalidAuthority,
+        "vault authority",
+    );
+
+    let signer_seeds = get_vault_pda_signer_seeds(ctx.accounts.config.key, &signer_bump);
+
+    // vault
+    // - must be the token account on the stake config account
+
+    require!(
+        ctx.accounts.vault.key == &config.vault,
+        StakeError::IncorrectVaultAccount,
+    );
+
+    let vault_data = ctx.accounts.vault.try_borrow_data()?;
+    // unpack to validate the mint
+    let vault = PodStateWithExtensions::<PodAccount>::unpack(&vault_data)?;
+
+    // mint
+    // - must match the stake vault mint
+
+    require!(
+        &vault.base.mint == ctx.accounts.mint.key,
+        StakeError::InvalidMint,
+        "mint"
+    );
+
+    // Update the stake amount on both stake and config accounts:
+    //
+    //   1. the amount slashed is taken from the active amount first;
+    //
+    //   2. in case the amount is not enough, the remaining is taken from
+    //      the inactive amount;
+    //
+    //   3. if the inactive stake is not enough, the remaining is taken
+    //      from the deactivating amount;
+    //
+    //   4. if still not enough, the remaining is ignored and the stake
+    //      account is left with 0 amount.
+
+    require!(
+        amount > 0,
+        StakeError::InvalidAmount,
+        "amount must be greater than 0"
+    );
+
+    let mut remaining_slash_amount = amount;
+
+    // First slash active stake.
+    let active_stake_to_slash = min(amount, stake.amount);
+    stake.amount = stake
+        .amount
+        .checked_sub(active_stake_to_slash)
+        .ok_or(ProgramError::ArithmeticOverflow)?;
+    remaining_slash_amount = remaining_slash_amount.saturating_sub(active_stake_to_slash);
+
+    // Then, if the slash amount has not been fulfilled, slash inactive stake
+    // (if there are inactive tokens).
+    if remaining_slash_amount > 0 && stake.inactive_amount > 0 {
+        let inactive_stake_to_slash = min(remaining_slash_amount, stake.inactive_amount);
+
+        msg!(
+            "+ slashing {} inactive stake tokens",
+            inactive_stake_to_slash
+        );
+
+        stake.inactive_amount = stake
+            .inactive_amount
+            .checked_sub(inactive_stake_to_slash)
+            .ok_or(ProgramError::ArithmeticOverflow)?;
+        remaining_slash_amount = remaining_slash_amount.saturating_sub(inactive_stake_to_slash);
+    }
+
+    // Then, if the slash amount has not been fulfilled, slash deactivating stake
+    // (if there are deactivating tokens).
```

Comment:
The deactivating token amount should be a portion of the currently active tokens, and not a separate amount. So if you've already slashed all of the active amount, that would also take the deactivating amount.

In fact, there should be an additional bit to make sure that deactivating amount isn't greater than amount after the slash, and should be set to `min(stake.deactivating_amount, stake.amount)`

Reply:
True, good point, makes no sense since the `amount` only gets updated when they move from `deactivating` to `inactive`. I will fix it.